pkg/operator/mon/mon.go

```diff
@@ -36,187 +39,399 @@ const (
+}
+
+func (c *Cluster) CheckHealth() error {
+	// update the config map if the pod ips changed
+	// must retry since during startup of pods they might take some time to initialize
+	k8sutil.Retry(time.Duration(c.retryDelay)*time.Second, c.maxRetries, func() (bool, error) {
+		err := c.updateConfigMapIfPodIPsChanged()
+		if err != nil {
+			logger.Infof("unable to check on mon pods. %+v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	// connect to the mons
+	context := &clusterd.Context{ConfigDir: c.configDir}
+	conn, err := mon.ConnectToClusterAsAdmin(context, c.factory, c.clusterInfo)
 	if err != nil {
-		return nil, fmt.Errorf("failed to start mon pods. %+v", err)
+		return fmt.Errorf("cannot connect to cluster. %+v", err)
+	}
+	defer conn.Shutdown()
+
+	// get the status and check for quorum
+	status, err := client.GetMonStatus(conn)
+	if err != nil {
+		return fmt.Errorf("failed to get mon status. %+v", err)
+	}
+
+	for _, monitor := range status.MonMap.Mons {
+		inQuorum := monInQuorum(monitor, status.Quorum)
+		if inQuorum {
+			logger.Debugf("mon %s found in quorum", monitor.Name)
+		} else {
+			logger.Infof("mon %s NOT found in quroum. %+v", monitor.Name, status.Quorum)
+
+			err = c.failoverMon(conn, monitor.Name)
+			if err != nil {
+				logger.Errorf("failed to failover mon %s. %+v", monitor.Name, err)
+			}
+		}
 	}
 
-	return clusterInfo, nil
+	return nil
+}
+
+func (c *Cluster) updateConfigMapIfPodIPsChanged() error {
+	pods, err := c.getPods()
+	if err != nil {
+		return fmt.Errorf("failed to check if pod ips changed. %+v", err)
+	}
+	if len(pods.Items) == 0 {
+		return fmt.Errorf("where are the mon pods?")
+	}
+
+	logger.Debugf("there are %d mon pods. checking the podIPs.", len(pods.Items))
+	updated := false
+	for _, pod := range pods.Items {
+		if pod.Status.PodIP == "" {
+			return fmt.Errorf("no podIP for mon %s", pod.Name)
+		}
+		if pod.Status.Phase != v1.PodRunning {
+			return fmt.Errorf("pod %s is not running. phase=%v", pod.Name, pod.Status.Phase)
+		}
+		m := mon.ToCephMon(pod.Name, pod.Status.PodIP)
+		existing, ok := c.clusterInfo.Monitors[pod.Name]
+		if !ok || existing.Endpoint == m.Endpoint {
+			// the endpoint does not need to be updated
+			logger.Debugf("Did not need to update pod %s with endpoint %+v. ok=%t, m=%+v", pod.Name, pod.Status.PodIP, ok, m)
+			continue
+		}
+
+		logger.Infof("updating mon %s endpoint from %s to %s", pod.Name, existing.Endpoint, m.Endpoint)
+		c.clusterInfo.Monitors[pod.Name] = m
+		updated = true
+	}
+
+	if updated {
+		return c.saveMonConfig()
+	}
+
+	logger.Debugf("no update to mon pod ips")
+	return nil
+}
+
+func monInQuorum(monitor client.MonMapEntry, quorum []int) bool {
+	for _, rank := range quorum {
+		if rank == monitor.Rank {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Cluster) failoverMon(conn client.Connection, name string) error {
+	logger.Infof("Failing over monitor %s", name)
+
+	// Start a new monitor
+	c.maxMonID++
```

Comment:
let's think further if we can get rid of this book keeping we are doing for the max (next) mon ID.  it'd be nice if that could be derived from the monMap we already have in the caller of this function.